src/routes/publish.ts

```diff
@@ -8,196 +8,602 @@ import { StatsWalesApi } from '../services/stats-wales-api';
+    // from errors in a previous attempt to just starting a new dataset.
+    // So lets clean up any remaining session data.
+    cleanupSession(req);
+    res.render('publish/start', { errors });
+});
+
+publish.get('/title', (req: AuthedRequest, res: Response) => {
+    res.render('publish/title', { errors: req.session.errors });
+});
+
+publish.post('/title', upload.none(), (req: AuthedRequest, res: Response) => {
+    if (!req.body?.title) {
+        logger.error('The user failed to supply a title in the request');
+        res.status(400);
+        res.render(
+            'publish/title',
+            {
+                errors: generateViewErrors(undefined, 500, [generateError('title', 'errors.title.missing', {})])
+            });
+        return;
+    }
+    req.session.currentTitle = req.body.title;
+    req.session.save();
+    const lang = req.i18n.language;
+    res.redirect(
+        `/${lang}/${req.i18n.t('routes.publish.start', { lng: lang })}/${req.i18n.t('routes.publish.upload', { lng: lang })}`
+    );
+});
+
+publish.get('/upload', (req: AuthedRequest, res: Response) => {
+    const currentTitle = req.session.currentTitle;
+    const currentDataset = req.session.currentDataset;
+    if (!currentDataset && !currentTitle) {
+        logger.error('There is no title or currentDataset in the session.  Abandoning this create journey');
+        req.session.errors = generateViewErrors(undefined, 500, [generateError('title', 'errors.title.missing', {})]);
         req.session.save();
-        res.redirect(`/${lang}/publish/preview`);
+        const lang = req.i18n.language;
+        res.redirect(
+            `/${lang}/${req.i18n.t('routes.publish.start', { lng: lang })}/${t('routes.publish.title', { lng: lang })}`
+        );
+        return;
+    }
+    const title = currentDataset?.datasetInfo?.find((info) => info.language === req.i18n.language) || currentTitle;
+    res.render('publish/upload', { title });
+});
+
+publish.post('/upload', upload.single('csv'), async (req: AuthedRequest, res: Response) => {
+    if (req.session.currentDataset) {
+        logger.info('Dataset present... Amending existing Dataset');
+        await uploadNewFileToExistingDataset(req, res);
     } else {
-        res.status(400);
-        res.render('publish/upload', processedCSV);
+        logger.info('Creating a new dataset');
+        await createNewDataset(req, res);
     }
 });
 
 publish.get('/preview', async (req: AuthedRequest, res: Response) => {
     const lang = req.i18n.language;
     const statsWalesApi = new StatsWalesApi(lang, req.jwt);
 
-    const dataset = req.session.currentDataset;
-    if (!dataset) {
-        logger.error('No dataset in session');
-        res.redirect(`/${req.i18n.language}/publish/title`);
+    const currentDataset = checkCurrentDataset(req, res);
+    if (!currentDataset) {
+        return;
+    }
+
+    const currentRevision = checkCurrentRevision(req, res);
+    if (!currentRevision) {
+        return;
```

Comment:
It typically returns a 500 error and redirects the user back to the start of the publish journey.  This is because something has gone wrong with their session.  Also again typescript so you need to make sure the thing exists before you can use it.